Mindbox.Data.Linq/Mapping/AttributedMetaModel.cs

```diff
@@ -181,18 +179,34 @@ public override MetaFunction GetFunction(MethodInfo method)
 	        if (method == null)
 		        throw Error.ArgumentNull("method");
 
-	        InitFunctions();
-            MetaFunction function;
-            metaFunctions.TryGetValue(new MetaPosition(method), out function);
-            return function;
-        }
+			var key = new MetaPosition(method);
 
-        public override IEnumerable<MetaFunction> GetFunctions() 
-		{
-            InitFunctions();
-            return metaFunctions.Values.ToList().AsReadOnly();
-        }
+			var function = metaFunctions.GetOrAdd(key, mp =>
+			{
+				if (IsUserFunction(method))
+				{
+					// Added this constraint because XML mapping model didn't support 
+					// mapping sprocs to generic method.
+					// The attribute mapping model was, however, able to support it. This check is for parity between 
+					// the two models.
+					if (method.IsGenericMethodDefinition)
+						throw Error.InvalidUseOfGenericMethodAsMappedFunction(method.Name);
+
+					var metaFunction = new AttributedMetaFunction(this, method);
+
+					// pre-set all known function result types into metaType map
```

Comment:
Isn't it too late to pre-set them here? This code no longer runs when during the initialization. Will it work without those pre-set types?

Reply:
Is there any difference in pre-set them or not? I don't think I understand why they want to initialize meta model with user defined functions return types, because you can add types to the meta model at any time, not only during initialization (just because l2s doesn't have any initialization phase). I think, this is just some kind of optimization.

Reply:
I mean that in `InitStaticTables` you can see some code, that initializes metatypes from types of first generic parameters of instance fields of type `Table<>` in `DataContext` class. But we don't use this feature, we don't have any fields in `MindboxDataContext` of type `Table<Customer>`, and everything works fine, so I think this is just an optimizaiton.

Reply:
If it's indeed just an optimization, it's probably no longer relevant when not during initialization. So it's better to remove it then.